design-docs/001-virtual-chunk-improvements.md

```diff
@@ -0,0 +1,163 @@
+  * The container template is expanded with the `template_args` in the virtual chunks. If there are no enough template arguments or some are None, they are replaced by the default template arguments in the container. If there are too many, last ones are ignored.
+  * The chunk is fetched from the results URL.
+
+### Writing virtual chunks
+
+* Users must first update the repo configuration to include the needed `VirtualChunkContainers`.
+* When the user writes a virtual chunk ref, they pass in the `VirtualChunkRef` the index for the container and the template expansion arguments. An error is produced if the container is not found.
+* This is a bit inconvenient, having to declare the container before being able to write. An alternative would be to automatically create a container that includes only the bucket name as template. On writes we try to match all templates against the url, and use the first one that matches, if non does, we create a new container. We could add that as a new function in the future if it's not required on first version.
+
+## Check requirements are satisfied
+
+1. > Virtual chunks can be stored in all supported object stores
+
+   Now the virtual chunk containers inform what is the platform, and we can create the appropriate object store client at read time. These clients are cached for performance.
+2. > Multiple credentials sets for virtual chunk resolution
+
+    Now the user initializes a Repo instance by passing, if needed, a credential set for each virtual chunk container. If a chunk is read for which there are no credentials, we default to a common set of credentials for virtual chunks, and ultimately, to the Storage credentials if on the same platform.
+3. > Optimized manifests where the same prefix doesn't need to be repeated for every virtual chunk
+  
+    For large prefixes, the template + arguments creates some savings. For example, a template for virtual zarr like `some-long-bucket-name/with-some-even-longer-prefix/and-more-nesting/some-long-array-name-for-the-dataset/c/{}/{}/{}`.
+   On the other hand, the template arguments vector has a 24 character overhead, on top of the values themselves. So it's not a lot of savings for smaller prefixes.
+5. > Be able to know all the "places" pointed by a repo, without having to load every manifest in every version
+
+    This can now be retrieved directly from the repository config. It's basically the set of virtual chunk containers (or a subset at least)
+
+6. > Object level change detection (not serving a virtual chunk if the containing file changed)
+
+    We optionally store the last-modified timestamp of the object containing each chunk. This is not optimal, a hash would be better, but it takes more space. Also not optimal, storing it per chunk, when many chunks come from the same object, but we don't have other easy place to store it if they are using containers that include more than one object. We could add more configuration to optimize this in the future.
```

Comment:
> but it takes more space

Take as many bytes as you feel necessary, It doesn't need to be a hash of the data, but a unique value from the store. 

and these are stored *per*file*, whereas there may be one or many chunks per file.